fabric-gamerule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/rule/EnumRule.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.gamerule.v1.rule;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.world.GameRules;
+
+import net.fabricmc.fabric.api.gamerule.v1.GameRuleRegistry;
+
+public class EnumRule<E extends Enum<E>> extends LiteralRule<EnumRule<E>> {
+	private static final Logger LOGGER = LogManager.getLogger(GameRuleRegistry.class);
+
+	private final Class<E> classType;
+	protected final Collection<E> supportedValues;
+	protected E value;
+
+	@Deprecated
+	public EnumRule(GameRules.RuleType<EnumRule<E>> type, E value, E[] supportedValues) {
+		this(type, value, Arrays.asList(supportedValues));
+	}
+
+	@Deprecated
+	public EnumRule(GameRules.RuleType<EnumRule<E>> type, E value, Collection<E> supportedValues) {
+		super(type);
+		this.classType = value.getDeclaringClass();
+		this.value = value;
```

Comment:
This should ensure the passed value is non-null

Reply:
`NullPointerException` due to null arguments is standard, although it could be thrown manually to provide context.

Reply:
Plus these constructors are deprecated since you should not be calling them. Unless we want to result to mixins into the api itself to expose the ctors,